File: playwright/pageFactory/ui/DynamicLoadingPage.ts
import type { Locator, Page } from '@playwright/test';
import { BasePage } from './BasePage';

export class DynamicLoadingPage extends BasePage {
	readonly heading: Locator;
	readonly subHeading: Locator;
	readonly startButton: Locator;
	readonly loading: Locator;
	readonly finishText: Locator;

	constructor(page: Page) {
		super(page);

		this.heading = page.locator('h3');
		this.subHeading = page.locator('h3 ~ h4');
		this.startButton = page.locator('#start button');
		this.loading = page.locator('#loading');
		this.finishText = page.locator('#finish');
	}

	async goto(): Promise<void> {
		await this.page.goto('/dynamic_loading');
	}

	async gotoExample1(): Promise<void> {
		await this.page.goto('/dynamic_loading/1');
	}

	async gotoExample2(): Promise<void> {
		await this.page.goto('/dynamic_loading/2');
	}

	async clickStart(): Promise<void> {
		await this.startButton.click();
	}
}

File: framework/src/main/java/ra/hul/framework/web/pages/DynamicLoadingPage.java
package ra.hul.framework.web.pages;

import io.qameta.allure.Step;
import org.openqa.selenium.By;
import ra.hul.framework.core.constants.Endpoints;
import ra.hul.framework.web.utils.WaitUtils;

public class DynamicLoadingPage extends BasePage {

    private final By startButton = By.cssSelector("#start button");
    private final By loadingIndicator = By.id("loading");
    private final By finishText = By.cssSelector("#finish h4");

    @Step("Open dynamic loading example {exampleNumber}")
    public DynamicLoadingPage open(int exampleNumber) {
        navigateToPath(Endpoints.DYNAMIC_LOADING + "/" + exampleNumber);
        return this;
    }

    @Step("Click start button")
    public DynamicLoadingPage clickStart() {
        click(startButton);
        return this;
    }

    @Step("Wait for loading to complete and get result")
    public String waitForResult() {
        WaitUtils.waitForInvisible(loadingIndicator, 30);
        return getText(finishText);
    }

    @Override
    public boolean isLoaded() {
        return isDisplayed(startButton);
    }
}
